### ISSUE TITLE
EventStore hangs / TCP connectivity errors / Duplicate events on CatchUp

### ISSUE THREAD

**Issue by haf** [EventStore/EventStore#1732]
I've finally isolated our issues with EventStore, reporting findings here as I debug the client/server.

We have lot of CatchUp subscriptions running. Our code assumes the don't duplicate events, but they are, and here's how that shows as the TCP connection state.

At some point, in this repro at 11:30:22 the TCP connection dies.

![screen shot 2018-09-21 at 19 29 34](https://user-images.githubusercontent.com/193115/45896533-aee8aa00-bdd4-11e8-8af5-4b83891186d6.png)

The server doesn't care about that, doesn't log the connection dying:

![screen shot 2018-09-21 at 19 29 52](https://user-images.githubusercontent.com/193115/45896550-c0ca4d00-bdd4-11e8-9c17-4b334957d5e7.png)

However, something's wrong with the server, and it ends up failing its health checks in k8s and gets restarted 2 minutes later:

![screen shot 2018-09-21 at 19 30 34](https://user-images.githubusercontent.com/193115/45896579-e0617580-bdd4-11e8-8ad1-a0fbf2327b58.png)

During that the duration of those two minutes, the system is down, because we can't receive events.

You can see the client lib heroically trying to reconnect, until the server gets rebooted.

![screen shot 2018-09-21 at 19 31 57](https://user-images.githubusercontent.com/193115/45896644-0edf5080-bdd5-11e8-81b6-d91c772e4652.png)

At this point, the catch-up subscriptions are hosed and some, or all of them start receiving duplicate events, which they act on, and send into ES on another stream (in our system):

![screen shot 2018-09-21 at 19 32 42](https://user-images.githubusercontent.com/193115/45896706-359d8700-bdd5-11e8-94a6-d9b10ab30771.png)

Now, since we track our subscription positions with the above stream, the next time our app reboots and reads from where it should start consuming, that position will be *n where n is the duplication factor.

This means we don't get events for a while, until the magic position equal to when the duplicated position is, and then we'll start receiving events without first receiving their previous events, queue invariant failures en-masse.

So; root cause is that the server pod dies and stops receiving/sending TCP data. This should be classified as a **major bug**.

Second root cause is that the catch up subscriptions don't properly handle disconnects without duplicating inputs to their callbacks. This should be classified as a **major bug**.

Cascade effect is that our invariant that "subscriptions don't give us duplicate events, so we can do event-counting" in violated.

Cascade effect is that when we reboot we resubscribe from the wrong position.

I haven't discovered *why* event store's pod dies; but it's not the node that goes down, it's something else. I also haven't discovered why the subscriptions duplicate input. I should be able to work around by abandoning catch up subscriptions and polling instead and also keeping track of the event ids whose results I persist to trigger the idempotent write mechanism.

This client is the latest netcore client.

Another cause leading up to this failure: there's no documented way of monitoring resource demands of the ES server; so when they exhaust, I don't know what resource got exhausted. Could be classified a **minor documentation bug**.

Also, I don't know if it's the TCP lingering and then disconnecting when there are no events, or if it's something else.

**Summary**: the server dies and cascading failures in the client and lack of monitoring docs leads to application invariant failures.

# [Swiss cheese](https://en.wikipedia.org/wiki/Swiss_cheese_model)

## ODE

This should be an exception resulting in 503 Unavailable Service from a HTTP API or effectively "pause" current execution, allowing callers to resume when the TCP socket it back up.

```
api           | System.AggregateException: One or more errors occurred. (Cannot access a disposed object.
api           | Object name: 'ES-b97c607f-c302-4174-9966-a804ae697e1a'.)
api           | --- End of exception stack trace ---
api           | System.ObjectDisposedException: Cannot access a disposed object.
api           | Object name: 'ES-b97c607f-c302-4174-9966-a804ae697e1a'.
api           |   at EventStore.ClientAPI.Internal.EventStoreNodeConnection.ReadEventAsync(String stream, Int64 eventNumber, Boolean resolveLinkTos, UserCredentials userCredentials) in /libs/EventStore.Client/Internal/EventStoreNodeConnection.cs:line 189
```

This happens WITH "KeepReconnecting" enabled, and causes the ES connection to indefinitely go into a failed state.

<img width="1221" alt="image" src="https://user-images.githubusercontent.com/193115/45943966-c39d8b80-bfe8-11e8-9e68-e696fd5e9df8.png">

This was mitigated by crashing the app any time "error occurred" happens. The mitigation leads to potential lost requests and harsher requirements for idempotence throughout  (requires: all POST requests have RequestId and are retries, all commands are identified by their request ids). This is an expensive mitigation, dev-time wise. It also leads to unavailability of our public service and should be classified as a **major bug**.

## Authentication failure

When ES server is not up and running and/or crashing, sometimes it gives "authentication failure" despite correct credentials being used. Only in the `Message` property can callers read _"Authentication timed out."_ This comes after the server comes back up, but before the server is ready to authenticate (I presume). This is a red herring in this context, since the issue here is that the server dies that the client can't handle errors thereof.

<img width="1232" alt="image" src="https://user-images.githubusercontent.com/193115/45945203-03667200-bfed-11e8-9206-e6388133a960.png">

Since the configuration of retrying the connection stops after getting "authentication failure" despite that failure being a timeout (which happen, we're talking dist. systems here!), it violates the invariant that `KeepReconnecting` configurations keep reconnecting and should be classified as a **major bug**.

# Ideas

## Heartbeat

I've played around a lot with ES on docker on my laptop, with docker-compose. I can get into a state where events aren't flowing after sleeping the laptop and then starting it up again. It would be good if there was a heartbeat mechanism in the client to avoid TCP hangs like this; a mechanism that could be monitored (and with monitoring documented).

**Comment by haf** [comment]
<img width="1241" alt="screen shot 2018-09-25 at 10 56 31" src="https://user-images.githubusercontent.com/193115/46003851-c0220700-c0b1-11e8-9542-1512211bce2c.png">

The EventStore server crashes about 2 times per 24 hours normally.

**Comment by gregoryyoung** [comment]
"Second root cause is that the catch up subscriptions don't properly handle disconnects without duplicating inputs to their callbacks. This should be classified as a major bug."

Catchupsubscriptions and their logic for getting/pushing data is quite basic in terms of logic (its literally just a checkpoint) and has been around for many years without others having issues. Maybe you can describe more or provide a sample case?

On a side note a catchupsubscription can *always* duplicate events up to whatever checkpoint interval is used on restarts etc. I have not heard of any case of duplicated events over a catchupsubscription outside of where its *known* that they can be duplicated. Perfect example you receive event 3 over subscription, you write to some db, you go to update checkpoint *power pull* when it comes back up event 3 will show up again as it was never checkpointed.

**Comment by haf** [comment]
> Catchupsubscriptions and their logic for getting/pushing data is quite basic in terms of logic (its literally just a checkpoint) and has been around for many years without others having issues. Maybe you can describe more or provide a sample case?

I'd love to, but I can't reproduce this besides waiting for the server to hang; only then does it happen. We can't crash it with stress tests otherwise, and it's not the client side that must hang; it's the server side. Perhaps this can help us in a few months when we've gotten around to building fault generation into the testing: https://istio.io/docs/tasks/traffic-management/fault-injection/ If the server would automatically log bad contextual conditions, such as out of memory, out of IO ops, out of handles/ulimit errors, then perhaps the logs would tell us under what conditions we can reproduce the error quicker (than the 12-16 hours it takes now) Simply put; I'm not good enough of a developer to reproduce this right now.

I do know replay is possible, and we have those checks peppered all over the software indeed, which is part of why it's been so hard to find this problem; a lot of the code handles it well. I'm sure I have places I've forgotten replay protection; like when I interact with external APIs that don't take request Ids; in those cases I've reasoned a one-in-a-million replay would be alright; but this happens every day for us.

**Comment by haf** [comment]
It's now been two weeks since this public report and more than six months since the initial contact with EventStore. This is a polite inquiry into how you are doing on handling the above mentioned issues?

**Comment by haf** [comment]
Related:

 - https://groups.google.com/forum/#!topic/event-store/z0RWBnBesoU — KeepReconnecting was tried here but shown not to work
 - https://github.com/EventStore/EventStore/issues/1584 — solution doesn't guarantee that the next time Connect is called, the next instance is not also disposed; also solution goes against docs ["to treat ES Connection as a singleton"](https://groups.google.com/forum/#!searchin/event-store/reconnect%7Csort:date/event-store/Ct5gAtNj4cE/z_uTnQqLOEIJ) and goes against API "KeepReconnecting"

**Comment by haf** [comment]
It's now been over two months since this was reported? What are you guys doing?

**Comment by jageall** [comment]
fixed by #2367 . auto reconnect was using the wrong position. catchup subscriptions were the only subscription doing that and caused all sorts of weird edge cases.  SubscriptionDropped handler is now called when the connection fails and it is expected that the handler reconnects using the checkpoint that it processed to. #2367

**Comment by haf** [comment]
What a horrendous turn around time! For such a core feature! 

Two years and we even bought the pro version/a support package to get support for this bug. Which you later cancelled because “it wasn’t free consulting hours”. 

All along you just ignored it and answered with an attitude to our problems. 👎 

Wow. You should be ashamed.
